src/server/chunk_server/chunk_server_file_service_impl.cc

```diff
@@ -131,8 +142,120 @@ grpc::Status ChunkServerFileServiceImpl::ReadFileChunk(
+
+  LOG(INFO) << "Checking if we have lease on chunk handle: "
+            << request_header.chunk_handle();
+
+  if (!chunk_server_impl_->HasWriteLease(request_header.chunk_handle())) {
+    // Doesn't have write lease
+    LOG(ERROR) << "Don't have write lease for chunk handle: "
+               << request_header.chunk_handle();
+    reply->set_status(FileChunkMutationStatus::FAILED_NOT_LEASE_HOLDER);
+    return grpc_status;
+  }
+
+  LOG(INFO) << "There is a valid lease on chunk handle: "
+            << request_header.chunk_handle();
+
+  // We currently do one write at a time and send apply mutation for just this
+  // write. The WriteFileChunkReply status is set as FileChunkMutationStatus::OK
+  // if this internal write succeeds. Because it can fail at other replicas, we
+  // add the status of each replica apply mutation to the replica_status.
+
+  // Get data from cache and try to make the local write
+  grpc_status = WriteFileChunkInternal(request_header, reply);
+
+  if (reply->status() != FileChunkMutationStatus::OK) {
+    // Write failed
+    return grpc_status;
+  }
+
+  // Write successful
+  // Send to other replicas to apply this write
+  LOG(INFO) << "Now sending apply mutation requests to "
+            << request->replica_locations_size()
+            << " replica(s) in parallel for file chunk: "
+            << request_header.chunk_handle();
+
+  // Send the requests in parallel
+  std::vector<
+      std::future<std::pair<std::string, StatusOr<ApplyMutationsReply>>>>
+      apply_mutation_results;
+
+  for (int replica = 0; replica < request->replica_locations_size();
+       ++replica) {
+    apply_mutation_results.push_back(
+        std::async(std::launch::async, [&, replica]() {
+          auto& replica_location = request->replica_locations(replica);
+
+          std::string hostname = replica_location.server_hostname();
+          if (chunk_server_impl_->ResolveHostname()) {
+            hostname = chunk_server_impl_->GetConfigManager()->ResolveHostname(
+                hostname);
+          }
+
+          auto server_address =
+              absl::StrCat(hostname, ":", replica_location.server_port());
+
+          auto client =
+              chunk_server_impl_->GetChunkServerProtocolClient(server_address);
+
+          ApplyMutationsRequest apply_mutation_request;
+          *apply_mutation_request.add_headers() = request_header;
+
+          LOG(INFO)
+              << "Sending apply mutation request to replica chunk server: "
+              << server_address
+              << " for file chunk: " << request_header.chunk_handle();
+
+          auto apply_mutation_reply =
+              client->SendRequest(apply_mutation_request);
+
+          return std::pair<std::string, StatusOr<ApplyMutationsReply>>(
+              server_address, apply_mutation_reply);
+        }));
+  }
+
+  // Wait for the apply muatation replies
+  for (int replica = 0; replica < request->replica_locations_size();
```

Comment:
Technically, we can save one rpc for the master to apply chunk write to itself directly right?

Reply:
Hmm, you mean primary not master? My assumption was you don't include the primary server in the replica location. From the comment in the grpc: 

```
// The replicas that the client has already sent the data to, so the primary
// chunk server knows who to forward the serialized concurrent writes to.
repeated protos.ChunkServerLocation replica_locations = 2;
```

In your client request, do you also include the primary?

Reply:
Discussed with Xi, primary isn't included in replica_locations.